ccm_web/client/src/pages/GradebookCanvas.tsx

```diff
@@ -0,0 +1,395 @@
+  }
+
+  const handleRowLevelInvalidationError = (errorMessage: JSX.Element[], invalidations: GradebookRowInvalidation[]): void => {
+    setPageState({ state: GradebookCanvasPageState.InvalidUpload, invalidations: invalidations, errorMessage: errorMessage })
+  }
+
+  const handleParseSuccess = (grades: GradebookRecord[]): void => {
+    setPageState({ state: GradebookCanvasPageState.Confirm, grades: grades })
+    setCSVtoDownload(grades)
+  }
+
+  const handleParseComplete = (results: ParseResult<GradebookRecord>): void => {
+    const data = results.data.slice(1) // The first row is possible scores
+
+    if (data[0]['Final Grade'] === undefined) {
+      handleNoLetterGradesError()
+      return
+    }
+
+    let invalidations: GradebookRowInvalidation[] = []
+
+    const gradeMismatchValidator = new CurrentAndFinalGradeMatchGradebookValidator()
+
+    data.forEach(record => {
+      invalidations = invalidations.concat(gradeMismatchValidator.validate(record, data.indexOf(record) + 1))
+    })
+
+    if (invalidations.length > 0) {
+      handleRowLevelInvalidationError([<div key='0'>There are blank cells in the gradebook. Please enter 0 or EX (for excused) for any blank cells in the gradebook and export a new CSV file.</div>], invalidations)
+    } else {
+      handleParseSuccess(data)
+    }
+  }
+
+  const renderUploadHeader = (): JSX.Element => {
+    return <span>
+      <Typography variant='h6'>Upload your CSV File</Typography>
+      <Typography>The CSV file will be formatted by trimming out nonessential columns.</Typography>
+      <br/>
+      <Typography><strong>Requirement needed:</strong> <Link href='#'>Grading Scheme in settings</Link> needs to be check marked for letter grade to appear in the CSV file.</Typography>
+    </span>
+  }
+
+  const renderFileUpload = (): JSX.Element => {
+    return <span>
+      <Grid container>
+        <Grid item xs={12}>
+          <FileUpload onUploadComplete={uploadComplete}></FileUpload>
+        </Grid>
+      </Grid>
+    </span>
+  }
+
+  const renderUpload = (): JSX.Element => {
+    return <span>
+      {renderUploadHeader()}
+      <br/>
+      {renderFileUpload()}
+    </span>
+  }
+
+  const renderUploadAgainButton = (): JSX.Element => {
+    return <Button color='primary' component="span" onClick={() => resetPageState()}>Upload again</Button>
+  }
+  const renderRowLevelErrors = (invalidations: GradebookRowInvalidation[]): JSX.Element => {
+    return (
+      <div>
+        {renderCSVFileName()}
+        <Grid container justify='flex-start'>
+          <Box clone order={{ xs: 2, sm: 1 }}>
+            <Grid item xs={12} sm={9} className={rowLevelErrorClasses.table} >
+              <ValidationErrorTable invalidations={invalidations} />
+            </Grid>
+          </Box>
+          <Box clone order={{ xs: 1, sm: 2 }}>
+            <Grid item xs={12} sm={3} className={rowLevelErrorClasses.dialog}>
+              <Paper role='alert' >
```

Comment:
Gonzolo pointed out that in a SPA you should use appropriate roles when new areas of attention appear.